playbooks/kubernetes/app-resource-tuning.md

```diff
@@ -0,0 +1,111 @@
+Kubernetes has knobs that allow you to allocate cpu/memory for an app.
+
+These settings are configured in k8s manifests (Hokusai staging/production.yml), in each app's repo. Hokusai templates come with example values for these settings:
+
+https://github.com/artsy/artsy-hokusai-templates
+
+However, we should tune them per app. This doc provides some guidelines on tuning.
+
+
+# Container CPU/Memory Request/Limit.
+
+https://kubernetes.io/docs/tasks/configure-pod-container/assign-cpu-resource/
+
+https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/
+
+These parameters are set per container. `requests` means how much of a resource will be reserved for the container. `limits` means the maximum amount of a resource that the container is allowed to use. If an app hits CPU limit, it will get throttled and slows down but it won't terminate. If it hits memory limit, it will likely terminate with out-of-memory error.
+
+CPU values are expressed in "number of CPU's". For example, 0.2 means 0.2 cpus. The value can also be expressed in units of milli cpus. So 200m is also 0.2 cpus. We prefer to use milli cpu units.
+
+For memory, the value is in bytes.
+
+For a pod that has a web container as well as an nginx container, our practice is that we set these parameters for the web container only.
+
+```
+---
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: example-web
+  namespace: default
+spec:
+...
+  template:
+...
+    spec:
+      containers:
+        - name: example-web
+
+          resources:
+            requests:
+              cpu: ???
+              memory: ???
+            limits:
+              cpu: ???
+              memory: ???
+```
+
+## General Recommendations
+These are very general recommendations, as CPU and memory usage are both hard to gauge. Some apps such as Pulse use a lot of CPU but only momentarily. For memory, some apps such as Gemini try to use more and more until they hit OOM and crash.
+
+### CPU Request
+- Set it to as much as the app (pod) actually uses most of the time. Find that out by monitoring usage over weeks (see below). Beware that utilization might be constrained by HPA as discussed below.
+- Set it to at least 200m even if the app uses less than that.
+- Cap it at 1 cpu, but if the app can take advantage of more than 1 cpu, give it more but no more than 2. The more CPU's requested, the harder it is for k8s to find room to fit the pod.
+- For apps that work efficiently with many instances, we can request less cpu per pod and use many pods.
+
+### CPU Limit
+- For critical apps such as Gravity, leave it, do not set it. The benefit is that when workload spikes, the pods can handle the load by using any idle CPU on the node, over what it requested.
+
+- For less critical apps, set it to 1, or 2 if CPU request is set to greater than 1.

```

Comment:
We didn't try and frame this in Kubernete's "Quality of Service" classes https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod either "burstable" or "guaranteed" but perhaps we should do so now and consider critical apps guaranteed - non critical apps burstable?

If we do so we can start thinking about utilizing the CPU manager (after a k8s upgrade) https://kubernetes.io/blog/2018/07/24/feature-highlight-cpu-manager/

Reply:
That's interesting. I am down for exploring that.